[LLD] Add CLASS syntax to SECTIONS

mysterymath opened the pull request:
This allows the input section matching algorithm to be separated from output section descriptions. This allows a group of sections to be assigned to multiple output sections, providing an explicit version of --enable-non-contiguous-regions's spilling that doesn't require altering global linker script matching behavior with a flag. It also makes the linker script language more expressive even if spilling is not intended, since input section matching can be done in a different order than sections are placed in an output section.

The implementation reuses the backend mechanism provided by --enable-non-contiguous-regions, so it has roughly similar semantics and limitations. In particular, sections cannot be spilled into or out of INSERT, OVERWRITE_SECTIONS, or /DISCARD/. The former two aren't intrinsic, so it may be possible to relax those restrictions later.

mysterymath committed:
[LLD] Add CLASS syntax to SECTIONS

This allows the input section matching algorithm to be separated from
output section descriptions. This allows a group of sections to be
assigned to multiple output sections, providing an explicit version of
--enable-non-contiguous-regions's spilling that doesn't require altering
global linker script matching behavior with a flag. It also makes the
linker script language more expressive even if spilling is not intended,
since input section matching can be done in a different order than
sections are placed in an output section.

The implementation reuses the backend mechanism provided by
--enable-non-contiguous-regions, so it has roughly similar semantics and
limitations. In particular, sections cannot be spilled into or out of
INSERT, OVERWRITE_SECTIONS, or /DISCARD/. The former two aren't
intrinsic, so it may be possible to relax those restrictions later.

smithp35 commented on lld/docs/ELF/linker_script.rst:
```diff
@@ -198,13 +198,52 @@ the current location to a max-page-size boundary, ensuring that the next
+
+  SECTIONS {
+    CLASS(c) { *(.rodata.earlier) }
+    .rodata { *(.rodata) CLASS(c) (*.rodata.later) }
```
While CLASS(C) { *(.rodata.earlier) } isn't going to clash with any OutputSection description as there is no `:` before the `{`. It is possible that `CLASS(C)` will clash with an existing linker script with an object file called `CLASS` and a section called `c`. We could accept the risk as I think that example is contrived, or perhaps use some different syntax for example `CLASS[c]`

MaskRay replied:
`CLASS` could be a file name, conflicting with existing script. But we can accept the risk, and users could use `"CLASS"(.text)` if they want to use `CLASS` as the file name...

smithp35 commented on lld/docs/ELF/linker_script.rst:
```diff
@@ -198,13 +198,52 @@ the current location to a max-page-size boundary, ensuring that the next
+::
+
+  SECTIONS {
+    CLASS(c) { *(.rodata.earlier) }
```
Can things like SORT and EXCLUDE_FILE be used in the input section descriptions for CLASS? SORT is interesting as the sections in the class could be ordered before attempting to place them in output sections.

smithp35 commented on lld/docs/ELF/linker_script.rst:
```diff
@@ -198,13 +198,52 @@ the current location to a max-page-size boundary, ensuring that the next
+
+::
+
+  SECTIONS {
```
I'm assuming that CLASS(c) is within SECTIONS and not hoisted out of it so that input sections matching earlier input section descriptions don't match a CLASS with wildcards. For example:
```
SECTIONS {
  .rodata.first { specific_file.o(.rodata.earlier) }
  CLASS(c) { *(.rodata.earlier) }
  .rodata { *(.rodata) CLASS(c) (*.rodata.later) }
}
```
Although that could be worked around by defining a separate CLASS for specific_file.o(.rodata.earlier).

My hoisted syntax example
```
CLASSES {
  c1 { specific_file.o(.rodata.earlier) }
  c2 { *(.rodata.earlier) }
}
```
SECTIONS {
  .rodata.first { CLASS(c1) }
  .rodata { *(.rodata) CLASS(c2) (*.rodata.later) }
}
```

smithp35 reviewed (commented):
I've not taken a look at the code in detail yet. Will hope to do that over the coming week. I've limited myself to commenting on some of the syntax and error messages.

I do like the idea of having ways of controlling a section match without needing a full output section.

One potential use case is matching `SHT_NOBITS` sections generated from `__attribute__((section(.bss.<name>)))` quite often these want to be placed after the `*.bss` input section description but `*.bss.*` will match every section generated this way.

mysterymath committed:
Address review comments

- Used consistent terminology for error when section classes are
  referenced before their definition.
- " for error when sections assigned to a class are never referenced
- Add test for class references in OVERLAY; fix parser.

smithp35 commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -490,99 +494,131 @@ static void sortInputSections(MutableArrayRef<InputSectionBase *> vec,
-    sizeAfterPrevSort = ret.size();
+    for (InputSectionDescription *isd : scd->sc.commands) {
+      for (InputSectionBase *sec : isd->sectionBases) {
+        if (sec->parent == &outCmd || !flagsMatch(sec))
```
In the non class case there is a helpful comment
```
// Skip if the section was already matched by a different input
// section description within this output section or class.
```
Could be worth a similar comment as I found it helpful when trying to remember out what sec->parent == &outCmd meant.

I'm wondering if it is worth some helper lambda's to name some of the tests. For example `alreadyAssignedToCommand` `for (sec->parent == &outCmd)` and `alreadyAssigned` for `(sec->parent && isa<OutputSection>(sec->parent)` some of these could be used in the non class reference part.

That's purely style preference though, and it may end up looking worse so feel free to ignore.

mysterymath replied:
Changed this to `alreadyAssignedToOutCmd`, but the lambda does help centralize the documentation for this, which makes the code a bit more compact.

smithp35 reviewed (commented):
Apologies for taking some time to respond.

I don't have too many comments on the code, following it through it looks like it would implement the specification in `linker_script.rst`. I think it may be worth making a few more tests to look at some feature combinations, but otherwise looks good to me.

mysterymath committed:
Address review feedback

- Add an explicit test that CLASS(a b) doesn't parse as a reference.
- Add SORT_BY_ALIGNMENT in class def to demonstrate more complex
  matching.
- Rename InputSectionDescription className to classRef to make it
  clearer that the class is being referenced, not defined.
- Define the interaction between --enable-non-contiguous-regions and
  CLASS and add tests.
- Add alreadyAssignedToOutCmd lambda to centralize comments.

smithp35 reviewed (approved):
Thanks for the updates. LGTM from my side. Will be worth waiting to see if MaskRay has any comments.

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,433 @@
+.balign 2
+.byte 6
+
+#--- matching.ld
```
Nit: `.ld` is not conventional in the testsuite. Older tests might use `.script` while newer ones use simpler `.lds` or `.t`

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,433 @@
+# MATCHING-NEXT: 020301cc 0605
+# MATCHING:      .rodata.d
+# MATCHING-NEXT: 04
+
```
We don't use two blank lines as separators...

The usual style when there are many split-file parts is:

```
## xxx
# RUN: ld.lld ...

# CHECK: ...
```

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,433 @@
+## An input section in a class spills to a later class ref when the region of
+## its first ref would overflow. The spill uses the alignment of the later ref.
+
+# RUN: ld.lld -T spill.ld spill.o -o spill
```
We usually place positive tests first, then negative tests.

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,377 @@
+
+#--- matching.ld
+SECTIONS {
+  CLASS(a) { *(.rodata.a) }
```
Perhaps add spaces to increase nuance `CLASS( cd)` `CLASS(ef )`

mysterymath replied:
Done.

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,433 @@
+#--- already-defined.ld
+SECTIONS {
+  CLASS(a) { *(.rodata.a) }
+  CLASS(a) { *(.rodata.b) }
```
Consider adding `CLASS(b)` to test how we print errors when there are multiple duplicates

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -490,99 +494,132 @@ static void sortInputSections(MutableArrayRef<InputSectionBase *> vec,
+          continue;
+        bool isSpill = sec->parent && isa<OutputSection>(sec->parent);
+        if (!sec->parent || (isSpill && outCmd.name == "/DISCARD/"))
+          error("section '" + sec->name + "' cannot spill from/to /DISCARD/");
```
Many `error` can be changed to `errorOrWarn` unless `--noinhibit-exec` would crash lld.

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,433 @@
+#--- from-discard.ld
+SECTIONS {
+  CLASS(class) { *(.two_byte_section) }
+  /DISCARD/ : { CLASS(class) }
```
We need a test that `/DISCARD/ : { CLASS(class) }` works if we don't use the spill semantics.

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,433 @@
+  CLASS(class) { *(.two_byte_section) }
+  .a : { *(.one_byte_section) }
+}
+SECTIONS { .b : { CLASS(class) } } INSERT AFTER .a;
```
We need a test that `SECTIONS { .... CLASS } INSERT` works if we don't use the spill semantics.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -490,99 +494,132 @@ static void sortInputSections(MutableArrayRef<InputSectionBase *> vec,
-    sizeAfterPrevSort = ret.size();
+    for (InputSectionDescription *isd : scd->sc.commands) {
+      for (InputSectionBase *sec : isd->sectionBases) {
+        if (alreadyAssignedToOutCmd(sec) || !flagsMatch(sec))
```
`alreadyAssignedToOutCmd(sec)` seems impossible?

mysterymath replied:
You can say `CLASS(c) CLASS(c)` in an output section description, but the second reference should have no effect, by comparison with `*(*) *(*)`.

MaskRay reviewed (changes requested):
.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -490,99 +494,132 @@ static void sortInputSections(MutableArrayRef<InputSectionBase *> vec,
-  // input order).
-  sortByPositionThenCommandLine(sizeAfterPrevSort, ret.size());
 
   // The flag --enable-non-contiguous-regions may cause sections to match an
```
The comment needs an update

mysterymath replied:
Done.

MaskRay commented on lld/docs/ELF/linker_script.rst:
```diff
@@ -198,13 +198,52 @@ the current location to a max-page-size boundary, ensuring that the next
+Section Classes
+~~~~~~~~~~~~~~~
+
+``SECTIONS`` commands can define classes of input sections:
```
`The CLASS keyword inside a SECTIONS command can define`

mysterymath replied:
Done; also removed unnecessary `can` tense.

MaskRay commented on lld/docs/ELF/linker_script.rst:
```diff
@@ -198,13 +198,52 @@ the current location to a max-page-size boundary, ensuring that the next
+  }
+
+Input sections that are assigned to a class cannot be matched by later
+wildcards, just as if they had been assigned to an earlier output section. If a
```
patterns, since they cannot be matched by non-wildcard patterns either

mysterymath replied:
Ah, nice catch. Also updated the wording for `--enable-non-contiguous-regions` to match.

MaskRay commented on lld/docs/ReleaseNotes.rst:
```diff
@@ -38,6 +38,13 @@ ELF Improvements
+  regions by automatically spilling to later class references if a region would
+  overflow. This reduces the toil of manually packing regions (typical for
+  embedded). It also makes full LTO feasible in such cases, since IR merging
+  currently prevents the linker script from referring to input files. (TODO: PR
```
This can be updated now.

mysterymath replied:
Done.

mysterymath committed:
Address review feedback

- `auto` over `const auto` for lambdas
- Inline `alreadyAssignedToOutCmd`
- Use `.lds` as linker script file extension in test
- Move RUN, CHECK and test descriptors into split files
- Add another duplicate class definition to test
- Remove dead `loc` variable
- Add missing braces to an `if` statement with braced `else`
- Remove "An error is reported when" from test descriptions
- Remove the 'd' from "SHF_MERGEd sections"
- Use `errorOrWarn` for errors that still allow an object file to be
  produced
- Add tests that a class reference can be used in /DISCARD/ or INSERT.
- Update spill comment.
- Tighten doc wording.
- Add PR number to release notes.

mysterymath committed:
Add spaces around select class refs

mysterymath committed:
Correct formatting

mysterymath commented:
@MaskRay, is there anything left for me to do on this one?

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,432 @@
+
+#--- matching.lds
+## CLASS definitions match sections in linker script order. The sections may be
+## placed in a different order. Classes may derive from one another. Class # #
```
trailing `# #`

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,432 @@
+  .rodata : {
+    *(.rodata.*)
+    INPUT_SECTION_FLAGS(SHF_EXECINSTR) CLASS( cd)
+    CLASS(a)
```
Perhaps use `CLASS(a)CLASS(ef )` to show that the space is unneeded.

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,432 @@
+} INSERT AFTER .rodata;
+
+# RUN: ld.lld -T matching.lds matching.o -o matching
+# RUN: llvm-readobj -x .rodata -x .rodata.d -x .rodata.h matching |\
```
`objdump -s` might be easier:)

`{{$}}` might be useful to show that the there is no unmatched string at the end like `0605{{$}}` `08{{$}}` below

MaskRay reviewed (commented):
Sorry for the delay. I am still reading through the patch.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -490,104 +494,135 @@ static void sortInputSections(MutableArrayRef<InputSectionBase *> vec,
+  } else {
+    SectionClassDesc *scd = script->sectionClasses.lookup(cmd->classRef);
+    if (!scd) {
+      errorOrWarn("undefined section class '" + cmd->classRef + "'");
```
untested

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -745,8 +832,8 @@ void LinkerScript::processSymbolAssignments() {
       addSymbol(assign);
-    else
-      for (SectionCommand *subCmd : cast<OutputDesc>(cmd)->osec.commands)
+    else if (auto *od = dyn_cast<OutputDesc>(cmd))
```
We often name `OutputDesc` `osd` (output section description)

MaskRay commented on lld/ELF/LinkerScript.h:
```diff
@@ -78,7 +80,8 @@ enum SectionsCommandKind {
   InputSectionKind,
-  ByteKind    // BYTE(expr), SHORT(expr), LONG(expr) or QUAD(expr)
+  ByteKind, // BYTE(expr), SHORT(expr), LONG(expr) or QUAD(expr)
+  ClassKind,
```
Add `// CLASS(class_name)`

MaskRay commented on lld/ELF/LinkerScript.h:
```diff
@@ -228,6 +234,10 @@ class InputSectionDescription : public SectionCommand {
   uint64_t withFlags;
   uint64_t withoutFlags;
+
+  // If present, input section matching uses class membership instead of file
```
If non-empty, this indicates a CLASS(class_name) reference, and filePat/sectionPatterns are not used.

Suggest to move this after sectionPatterns to make it clear they are mutual exclusive.

MaskRay commented on lld/ELF/LinkerScript.h:
```diff
@@ -413,6 +422,11 @@ class LinkerScript final {
+  // Named lists of input sections that can be collectively referenced in output
+  // section descriptions. Multiple references allow for sections to spill from
+  // one output section to another.
+  llvm::StringMap<SectionClassDesc *> sectionClasses;
```
`DenseMap<CachedHashStringRef, >` or `<StringRef, >`

This header otherwise doesn't use StringMap

MaskRay commented on lld/test/ELF/linkerscript/section-class.test:
```diff
@@ -0,0 +1,432 @@
+# REFERENCED-BEFORE-DEFINED-WARN: warning: section class 'a' referenced by '.rodata' before class definition
+
+#--- unreferenced.lds
+## An input section is bound to a section class but is not referenced by at
```
`.. is not referenced.`

It's unclear what `at least one output section` means.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -717,14 +752,66 @@ void LinkerScript::processSectionCommands() {
+    for (InputSectionDescription *isd : sc->sc.commands)
+      for (InputSectionBase *sec : isd->sectionBases)
+        if (sec->parent && isa<SectionClass>(sec->parent))
+          errorOrWarn("section '" + sec->name + "' assigned to class '" +
```
The test shows one error/warning, but there are many. `.rodata.a` isn't very useful without an input file name.
```
ld.lld: error: section '.rodata.a' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.b' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.c' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.d' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.e' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.f' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.g' assigned to class 'a' but unreferenced by any output section
ld.lld: error: section '.rodata.h' assigned to class 'a' but unreferenced by any output section
```

Perhaps we should report one single error: `section class 'a' is unreferenced`.

We don't need to mention matched input sections, since that's clear.

mysterymath replied:
Simplified the comment along these lines and made it only report once per class.

MaskRay commented on lld/ELF/LinkerScript.cpp:
```diff
@@ -717,14 +752,66 @@ void LinkerScript::processSectionCommands() {
+        isd->sectionBases =
+            computeInputSections(isd, ctx.inputSections, sc->sc);
+        for (InputSectionBase *s : isd->sectionBases) {
+          // Section classes with --enable-non-contiguous-regions may contain
```
This comment is not accurate. This code path is useful even if --enable-non-contiguous-regions is unused.

mysterymath replied:
I think the comment was probably confusing; I've reworded it to hopefully be clearer. It should be the case that the only way a section can be assigned to the class *and* have a different parent than the class is using `--enable-non-contiguous-regions`; otherwise, having a parent would put it out of scope for matching a class.

MaskRay commented on lld/ELF/OutputSections.h:
```diff
@@ -137,6 +137,25 @@ struct OutputDesc final : SectionCommand {
   }
 };
 
+// A list of input sections that can be referenced in output descriptions.
```
// This represents a `CLASS(class_name) { ... }` that can be referenced by output section descriptions. If referenced more than once, the sections can be spilled to the next reference like --enable-non-contiguous-regions.

mysterymath replied:
Done.

MaskRay commented on lld/ELF/ScriptParser.cpp:
```diff
@@ -585,6 +587,34 @@ SmallVector<SectionCommand *, 0> ScriptParser::readOverlay() {
+
+StringRef ScriptParser::readSectionClassName() {
+  expect("(");
+  StringRef name = next();
```
We could use `unquote` to allow quoted name or just reject quoted names.

mysterymath replied:
Added `unquote` to allow working around the file-named-CLASS issue.

mysterymath committed:
Address review comments

- Remove trailing hashes from test
- Joined test class references together to demonstrate spaces not required
- Migrate to llvm-objdump -s for content tests; add end of line checks
- Add test for missing open paren not being interpreted as a filename
- Add test for undefined section class
- Use abbreviation "osd" for OutputDesc variables
- Add comment for section class command enum
- Move classRef member variable earlier in InputSectionDescription
- Use DenseMap from CachedHashStringRef over StringMap
- Clarify undefined.lds test comment.
- Clarify comment for section class members with different parent
- Replace comment on SectionClass struct
- Use unquote for section names; add test

mysterymath committed:
clang-format

MaskRay commented on lld/ELF/ScriptParser.cpp:
```diff
@@ -611,6 +613,34 @@ SmallVector<SectionCommand *, 0> ScriptParser::readOverlay() {
+  if (!script->sectionClasses.insert({CachedHashStringRef(name), desc}).second)
+    setError("section class '" + name + "' already defined");
+  expect("{");
+  while (!errorCount() && !consume("}")) {
```
With #100493  , the loop pattern needs a change.

mysterymath replied:
Hmm, I had used `readOutputSectionDescription` as a reference for the parsing idioms, but that didn't change in #100493 . Is there a substantive difference between that and `readSectionClassDescription`, or was the loop in `readOutputSectionDescription` missed? It appears that there are still 15 instances of the `while(!errorCount() && !consume(...` idiom in ScriptParser.cpp, but it isn't obvious to me what the differences are between those changed and those that remain.

MaskRay replied:
I neglected `while (... consume("}"))` in #100493 , which might lead to bad diagnostics or dead loops.

I introduced `till()` in 2a89356d64a274f7171b45d8b7b05b6315a6f627 and refactored some `while (... consume("}"))` to use `till`. Use that instead?

mysterymath replied:
Done.

mysterymath committed:
Use till

MaskRay commented on lld/docs/ReleaseNotes.rst:
```diff
@@ -26,6 +26,14 @@ Non-comprehensive list of changes in this release
+  allows the linker to automatically pack the input sections into memory
+  regions by automatically spilling to later class references if a region would
+  overflow. This reduces the toil of manually packing regions (typical for
+  embedded). It also makes full LTO feasible in such cases, since IR merging
```
How is full LTO feasible? I would expect that brief items in release notes have some description in the patch descriotion.

mysterymath replied:
Thanks, this went out of date now that we've cut the release branch. Reworded.

mysterymath committed:
Update release notes

mysterymath committed:
Simplify unreferenced class error message and only report once

MaskRay commented:
Note: per consensus on https://discourse.llvm.org/t/rfc-github-pr-resolve-conversation-button/73178 , "Resolve comment" buttons are for the reviewers to click, not the author.

mysterymath commented:
> Note: per consensus on https://discourse.llvm.org/t/rfc-github-pr-resolve-conversation-button/73178 , "Resolve comment" buttons (perhaps except really trivial ones) are for the reviewers to click, not the author.

Thanks; I'd missed that there was an RFC on this.
